client/src/panel.ts

```diff
@@ -0,0 +1,240 @@
+import vscode from 'vscode'
+import { LanguageClient } from 'vscode-languageclient/node'
+import { ensureCatchPromise } from './utils'
+
+// Modified from https://github.com/microsoft/vscode/blob/main/extensions/markdown-language-features/src/util/dispose.ts
+/**
+ * A basic implementer of `DisposableSupplemental`, meant to be delegated to
+ * by more complex implementers of `DisposableSupplemental`.
+ */
+export class Disposer implements DisposableSupplemental {
+  private _disposed = false
+  private registeredDisposables: Disposable[] = []
+  private readonly _onDidDispose: vscode.EventEmitter<void> = new vscode.EventEmitter()
+
+  readonly onDidDispose: vscode.Event<void> = (...args) => {
+    return this._onDidDispose.event(...args)
+  }
+
+  /**
+   * Disposes all child disposables
+   */
+  private disposeAll(): void {
+    for (const disposable of this.registeredDisposables) {
+      disposable.dispose()
+    }
+    this.registeredDisposables = []
+  }
+
+  dispose(): void {
+    if (this._disposed) {
+      return
+    }
+    this._disposed = true
+    this._onDidDispose.fire()
+    this.disposeAll()
+  }
+
+  registerDisposable<T extends Disposable>(value: T): T {
+    if (this.disposed()) {
+      value.dispose()
+    } else {
+      this.registeredDisposables.push(value)
+    }
+    return value
+  }
+
+  disposed(): boolean {
+    return this._disposed
+  }
+}
+
+/**
+ * An abstract class to be inherited by webview panel components.
+ *
+ * Should greedily create an internal panel during construction. Implementers
+ * may include extra fields and disposables bound to each instance of
+ * inheriting types to create unique behaviors. Instances of inheriting types
+ * will have a two-way disposal binding with the internal panel, i.e. if the
+ * internal panel is disposed, this item will be disposed as a result, or if
+ * this item is disposed, the internal panel will be disposed as a result.
+ *
+ * Inheriting types must specify generic parameters representing the types of
+ * messages that are sent back and forth between the internal panel and the
+ * host. `InMessage` is the type of messages sent from the internal webview
+ * panel to the host. `OutMessage` is the type of messages sent from the host
+ * to the internal webview panel.
+ */
+export abstract class Panel<InMessage, OutMessage> implements DisposableSupplemental, Messageable<InMessage, OutMessage> {
+  protected readonly panel: vscode.WebviewPanel
+  private readonly disposer: DisposableSupplemental
+
+  constructor(initPanel: () => vscode.WebviewPanel) {
+    this.disposer = new Disposer()
+    this.panel = initPanel()
```

Comment:
I like having a common base for all panels. If we're immediately calling initPanel in the constructor why not make the constructor take a Panel instead of a function that returns a Panel?

:thinking: Also, (for another PR) since this is a base for all panels, would it make sense to perform all the CSP incantations here in a common place?

Reply:
Huh... I'm not sure why I made it a function that returns a Panel... Must've over-thought it. I'll change that.

As for CSP stuff, I'm a little hesitant to do that, since the Panel class is abstract and really doesn't care at all what inheritors do when it comes to content. CSP isn't always listed out right now, e.g. there's a loading state for cnxml preview, where I didn't bother using csp stuff, and the panel class doesn't stop me. Were you thinking of this being a helper of some sort, or a mandate that the panel class enforces (that all displayed content must specify a CSP)?

Reply:
Let's continue this discussion, but I'm gonna go ahead and merge this PR, since it's got the two approvals, and I've addressed most of the things. I suspect I may have to come back to anti-jitter behavior as well sometime soon as well.